events/admin.py

```diff
@@ -25,43 +25,54 @@ def save_model(self, request, obj, form, change):
         system_id = settings.SYSTEM_DATA_SOURCE_ID
         obj.data_source_id = system_id
         if not obj.id:
-            obj.id = generate_id(system_id)
+            if obj.origin_id:
+                obj.id = ':'.join([system_id, obj.origin_id])
+            else:
+                obj.id = generate_id(system_id)
         obj.origin_id = obj.id.split(':')[1]
 
         super().save_model(request, obj, form, change)
 
 
 class EventAdmin(AutoIdBaseAdmin, TranslationAdmin, VersionAdmin):
-    # TODO: location, publisher, keyword, audience, super_event fields with autocomplete
     # TODO: only allow user_editable editable fields
-    fields = ('name', 'short_description', 'description', 'location', 'location_extra_info', 'start_time', 'end_time',
-              'keywords', 'audience', 'publisher', 'provider', 'provider_contact_info', 'event_status', 'super_event',
-              'info_url', 'in_language')
+    fields = ('id', 'name', 'short_description', 'description', 'location', 'location_extra_info', 'start_time',
+              'end_time', 'keywords', 'audience', 'publisher', 'provider', 'provider_contact_info', 'event_status',
+              'super_event', 'info_url', 'in_language')
     search_fields = ('name', 'location__name')
     list_display = ('id', 'name', 'start_time', 'end_time', 'publisher', 'location')
     list_filter = ('data_source',)
     ordering = ('-last_modified_time',)
+    autocomplete_fields = ('location', 'keywords', 'audience', 'super_event', 'publisher')
+    readonly_fields = ('id',)
 
 
 admin.site.register(Event, EventAdmin)
 
 
 class KeywordAdmin(AutoIdBaseAdmin, TranslationAdmin, VersionAdmin):
-    # TODO: publisher field with autocomplete
     # TODO: only allow user_editable editable fields
-    fields = ('publisher', 'deprecated', 'name')
+    fields = ('id', 'publisher', 'deprecated', 'name')
     search_fields = ('name',)
     list_display = ('id', 'name', 'n_events')
     list_filter = ('data_source',)
     ordering = ('-n_events',)
+    autocomplete_fields = ('publisher', )
+    readonly_fields = ('id',)
 
 
 admin.site.register(Keyword, KeywordAdmin)
 
 
-class KeywordSetAdmin(BaseAdmin):
-    # TODO: keywords field with autocomplete
-    fields = ('data_source', 'origin_id', 'name', 'keywords', 'usage')
+class KeywordSetAdmin(AutoIdBaseAdmin):
+    fields = ('id', 'data_source', 'origin_id', 'name', 'keywords', 'usage')
+    autocomplete_fields = ('keywords', )
+
+    def get_readonly_fields(self, request, obj=None):
+        if obj:
+            return ['id', 'origin_id', 'data_source']
+        else:
+            return ['id']
```

Comment:
Always add `data_source` here, since we don't want to allow specifying it. All new objects generated in the admin should come from `SYSTEM_DATA_SOURCE_ID`.

Reply:
Done.